File: frontend/src/services/stock-adjustments.ts
import { http } from '@/services/http';
import type { PaginatedResponse } from '@/types/common';

export type StockAdjustmentType = 'increase' | 'decrease';

export interface StockAdjustment {
  id: number;
  product_id: number;
  warehouse_id: number;
  adjustment_type: StockAdjustmentType;
  quantity: number;
  reason: string;
  note?: string | null;
  adjustment_reference?: string | null;
  created_by?: number | null;
  created_at: string;
}

export interface StockAdjustmentCreateInput {
  product_id: number;
  warehouse_id: number;
  adjustment_type: StockAdjustmentType;
  quantity: number;
  reason: string;
  note?: string;
  adjustment_reference?: string;
  allow_negative?: boolean;
}

export interface StockAdjustmentListParams {
  skip?: number;
  limit?: number;
  product_id?: number;
  warehouse_id?: number;
  adjustment_type?: StockAdjustmentType;
  date_from?: string;
  date_to?: string;
}

export interface CurrentStockResponse {
  product_id: number;
  warehouse_id: number;
  quantity: number;
}

export function buildStockAdjustmentsQuery(params: StockAdjustmentListParams = {}): string {
  const query = new URLSearchParams();
  if (params.skip !== undefined) query.append('skip', String(params.skip));
  if (params.limit !== undefined) query.append('limit', String(params.limit));
  if (params.product_id !== undefined) query.append('product_id', String(params.product_id));
  if (params.warehouse_id !== undefined) query.append('warehouse_id', String(params.warehouse_id));
  if (params.adjustment_type) query.append('adjustment_type', params.adjustment_type);
  if (params.date_from) query.append('date_from', params.date_from);
  if (params.date_to) query.append('date_to', params.date_to);
  return query.toString();
}

export const stockAdjustmentApi = {
  async list(params: StockAdjustmentListParams = {}): Promise<PaginatedResponse<StockAdjustment>> {
    const query = buildStockAdjustmentsQuery(params);
    const url = query ? `/stock-adjustments?${query}` : '/stock-adjustments';
    return http.get(url);
  },

  async getById(adjustmentId: number): Promise<StockAdjustment> {
    return http.get(`/stock-adjustments/${adjustmentId}`);
  },

  async create(data: StockAdjustmentCreateInput): Promise<StockAdjustment> {
    return http.post('/stock-adjustments', data);
  },

  async getCurrentStock(productId: number, warehouseId: number): Promise<CurrentStockResponse> {
    return http.get(`/stock-adjustments/current-stock?product_id=${productId}&warehouse_id=${warehouseId}`);
  },
};

export async function getStockAdjustments(params: StockAdjustmentListParams = {}): Promise<PaginatedResponse<StockAdjustment>> {
  return stockAdjustmentApi.list(params);
}

export async function createStockAdjustment(data: StockAdjustmentCreateInput): Promise<StockAdjustment> {
  return stockAdjustmentApi.create(data);
}

File: backend/app/api/v1/endpoints/stock_adjustments.py
"""Stock adjustment endpoints."""

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import check_permission
from app.schemas.common import PaginationResponse
from app.schemas.stock_adjustment import (
    StockAdjustmentCreate,
    StockAdjustmentCurrentStockResponse,
    StockAdjustmentResponse,
)
from app.services.audit_service import AuditService
from app.services.stock_adjustment_service import StockAdjustmentService

router = APIRouter()


def adjustment_to_dict(adjustment: Any) -> dict[str, Any]:
    """Serialize adjustment payload for audit logging."""
    adjustment_type = adjustment.adjustment_type
    if hasattr(adjustment_type, "value"):
        adjustment_type = adjustment_type.value

    return {
        "id": adjustment.id,
        "product_id": adjustment.product_id,
        "warehouse_id": adjustment.warehouse_id,
        "adjustment_type": adjustment_type,
        "quantity": adjustment.quantity,
        "reason": adjustment.reason,
        "note": adjustment.note,
        "adjustment_reference": adjustment.adjustment_reference,
        "created_by": adjustment.created_by,
        "created_at": (
            adjustment.created_at.isoformat()
            if adjustment.created_at
            else None
        ),
    }


@router.get("", response_model=PaginationResponse[StockAdjustmentResponse])
async def list_stock_adjustments(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
    product_id: int | None = Query(None),
    warehouse_id: int | None = Query(None),
    adjustment_type: str | None = Query(None, pattern="^(increase|decrease)$"),
    date_from: str | None = Query(None),
    date_to: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("stock_adjustment:view")),
):
    """List stock adjustments with filters."""
    service = StockAdjustmentService(db)

    parsed_date_from = None
    parsed_date_to = None
    if date_from:
        from datetime import datetime

        parsed_date_from = datetime.fromisoformat(date_from)
    if date_to:
        from datetime import datetime

        parsed_date_to = datetime.fromisoformat(date_to)

    items, total = await service.get_multi(
        skip=skip,
        limit=limit,
        product_id=product_id,
        warehouse_id=warehouse_id,
        adjustment_type=adjustment_type,
        date_from=parsed_date_from,
        date_to=parsed_date_to,
    )

    total_pages = (total + limit - 1) // limit
    page = skip // limit + 1

    return PaginationResponse(
        items=items,
        total=total,
        page=page,
        page_size=limit,
        total_pages=total_pages,
    )


@router.get(
    "/current-stock", response_model=StockAdjustmentCurrentStockResponse
)
async def get_current_stock(
    product_id: int = Query(..., gt=0),
    warehouse_id: int = Query(..., gt=0),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("stock_adjustment:view")),
):
    """Get current stock for selected product/warehouse."""
    service = StockAdjustmentService(db)
    quantity = await service.get_current_stock(
        product_id=product_id, warehouse_id=warehouse_id
    )

    return StockAdjustmentCurrentStockResponse(
        product_id=product_id,
        warehouse_id=warehouse_id,
        quantity=quantity,
    )


@router.get("/{adjustment_id}", response_model=StockAdjustmentResponse)
async def get_stock_adjustment(
    adjustment_id: int,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("stock_adjustment:view")),
):
    """Get stock adjustment by ID."""
    service = StockAdjustmentService(db)
    adjustment = await service.get_by_id(adjustment_id)

    if not adjustment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Stock adjustment not found",
        )

    return adjustment


@router.post(
    "",
    response_model=StockAdjustmentResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_stock_adjustment(
    payload: StockAdjustmentCreate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("stock_adjustment:create")),
):
    """Create stock adjustment and post stock ledger movement."""
    service = StockAdjustmentService(db)
    audit_service = AuditService(db)

    if await service.has_reference_conflict(payload.adjustment_reference):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Adjustment reference already exists",
        )

    adjustment = await service.create_adjustment(
        payload=payload, actor_id=current_user.id
    )

    await audit_service.create_log(
        user_id=current_user.id,
        action="create",
        resource_type="stock_adjustment",
        resource_id=adjustment.id,
        description=f"Created stock adjustment {adjustment.id}",
        new_values=adjustment_to_dict(adjustment),
        ip_address=request.client.host if request.client else None,
        user_agent=request.headers.get("user-agent"),
    )

    return adjustment
